File: lingting-security/src/main/java/live/lingting/framework/security/resource/SecurityHolder.java
package live.lingting.framework.security.resource;

import live.lingting.framework.security.domain.SecurityScope;
import live.lingting.framework.thread.StackThreadLocal;

import java.util.Collections;
import java.util.Optional;
import java.util.Set;

/**
 * @author lingting 2023-03-29 20:29
 */
public final class SecurityHolder {

	static final StackThreadLocal<SecurityScope> LOCAL = new StackThreadLocal<>();

	private SecurityHolder() {throw new UnsupportedOperationException("This is a utility class and cannot be instantiated");}

	public static void put(SecurityScope scope) {
		LOCAL.put(scope);
	}

	public static void pop() {
		LOCAL.pop();
	}

	public static SecurityScope get() {
		return scope();
	}

	public static Optional<SecurityScope> option() {
		return scopeOption();
	}

	public static SecurityScope scope() {
		return LOCAL.get();
	}

	public static Optional<SecurityScope> scopeOption() {
		return Optional.ofNullable(scope());
	}

	public static String token() {
		return scopeOption().map(SecurityScope::getToken).orElse("");
	}

	public static String userId() {
		return scopeOption().map(SecurityScope::getUserId).orElse(null);
	}

	public static String tenantId() {
		return scopeOption().map(SecurityScope::getTenantId).orElse(null);
	}

	public static String username() {
		return scopeOption().map(SecurityScope::getUsername).orElse("");
	}

	public static String password() {
		return scopeOption().map(SecurityScope::getPassword).orElse("");
	}

	public static Set<String> roles() {
		return scopeOption().map(SecurityScope::getRoles).orElse(Collections.emptySet());
	}

	public static Set<String> permissions() {
		return scopeOption().map(SecurityScope::getPermissions).orElse(Collections.emptySet());
	}

}

File: lingting-security/src/main/kotlin/live/lingting/framework/security/resource/SecurityHolder.kt
package live.lingting.framework.security.resource

import java.util.Optional
import live.lingting.framework.context.StackContext
import live.lingting.framework.security.domain.SecurityScope

/**
 * @author lingting 2023-03-29 20:29
 */
object SecurityHolder {

    val LOCAL: StackContext<SecurityScope?> = StackContext()

    @JvmStatic
    fun put(scope: SecurityScope?) {
        LOCAL.push(scope)
    }

    @JvmStatic
    fun pop() {
        LOCAL.pop()
    }

    @JvmStatic
    fun get(): SecurityScope? {
        return scope()
    }

    @JvmStatic
    fun option(): Optional<SecurityScope> {
        return scopeOption()
    }

    @JvmStatic
    fun scope(): SecurityScope? {
        return LOCAL.peek()
    }

    @JvmStatic
    fun scopeOption(): Optional<SecurityScope> {
        return Optional.ofNullable(scope())
    }

    @JvmStatic
    fun authorization(): String {
        return scopeOption().map { it.authorization }.orElse("")
    }

    @JvmStatic
    fun userId(): String? {
        return scopeOption().map { it.userId }.orElse(null)
    }

    @JvmStatic
    fun tenantId(): String? {
        return scopeOption().map { it.tenantId }.orElse(null)
    }

    @JvmStatic
    fun username(): String {
        return scopeOption().map { it.username }.orElse("")
    }

    @JvmStatic
    fun password(): String {
        return scopeOption().map { it.password }.orElse("")
    }

    @JvmStatic
    fun roles(): Set<String> {
        return scopeOption().map { it.roles }.orElse(emptySet<String>())
    }

    @JvmStatic
    fun permissions(): Set<String> {
        return scopeOption().map { it.permissions }.orElse(emptySet<String>())
    }

}
